--- ../src-base/minecraft/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
@@ -15,9 +15,9 @@
 {
     private final AbstractAttributeMap attributeMap;
     private final IAttribute genericAttribute;
-    private final Map<Integer, Set<AttributeModifier>> mapByOperation = Maps.<Integer, Set<AttributeModifier>>newHashMap();
-    private final Map<String, Set<AttributeModifier>> mapByName = Maps.<String, Set<AttributeModifier>>newHashMap();
-    private final Map<UUID, AttributeModifier> mapByUUID = Maps.<UUID, AttributeModifier>newHashMap();
+    private final Map<Integer, Set<AttributeModifier>> mapByOperation = Maps.<Integer, Set<AttributeModifier>>newConcurrentMap();
+    private final Map<String, Set<AttributeModifier>> mapByName = Maps.<String, Set<AttributeModifier>>newConcurrentMap();
+    private final Map<UUID, AttributeModifier> mapByUUID = Maps.<UUID, AttributeModifier>newConcurrentMap();
     private double baseValue;
     private boolean needsUpdate = true;
     private double cachedValue;
@@ -85,7 +85,8 @@
     {
         if (this.getModifier(modifier.getID()) != null)
         {
-            throw new IllegalArgumentException("Modifier is already applied on this attribute!");
+            new IllegalArgumentException("Modifier is already applied on this attribute!").printStackTrace();
+            // throw new IllegalArgumentException("Modifier is already applied on this attribute!");
         }
         else
         {
